### ISSUE TITLE
Sequel Ace crashes attempting to view the tables in the information_schema db on MySQL v8

### ISSUE THREAD

**Issue by lphilps** [Sequel-Ace/Sequel-Ace#833]
- Sequel Ace Version: 3.1.1 Build 3016
- Sequel Ace Source: App Store
- macOS Version: 10.15.7
- Processor Type: Intel
- MySQL Version: 8.0.20
- macOS Localization: English

Is Issue Present in Latest Beta? Not according to the release notes associated with 3.2.0 Beta1

**Description**
Attempting to view the content of any table in the information_schema database causes Sequel Ace to crash, but only when connected to a MySQL v8 database.  Works fine when connected to a v5.7 db server.

**Steps To Reproduce**
1. Open Sequel Ace and connect to a MySQL v8 database.  I connected to a db running 8.0.20 via an ssh tunnel.
2. Make sure the Content view is active
3. In the database selector, select "information_schema"
4. Click on any table listed - I've tried COLUMNS, KEY_COLUMN_USAGE, and TABLES
5. Sequel Ace crashes with a Segmentation Fault.

**Expected Behaviour**
I expected to see the table contents.

**Related Issues**
None that I know of

**Additional Context**
[Crashlog.txt](https://github.com/Sequel-Ace/Sequel-Ace/files/5874628/Crashlog.txt)

**Comment by jamesstout** [comment]
Cause is this block: https://github.com/Sequel-Ace/Sequel-Ace/blob/69e2d39d907e695cd12575351fd9e673ce5ca43f/Source/Controllers/MainViewControllers/TableContent/SPTableContent.m#L774-L786

Specifically L:783

**Comment by jamesstout** [comment]
Hmm, the issue is, on version < 8.0.20, COLUMNS is a table. On 8.0.20 it's a complicated view.

**5.7.9**
```sql
CREATE TEMPORARY TABLE `COLUMNS` (
  `TABLE_CATALOG` varchar(512) NOT NULL DEFAULT '',
  `TABLE_SCHEMA` varchar(64) NOT NULL DEFAULT '',
  `TABLE_NAME` varchar(64) NOT NULL DEFAULT '',
  `COLUMN_NAME` varchar(64) NOT NULL DEFAULT '',
  `ORDINAL_POSITION` bigint(21) unsigned NOT NULL DEFAULT '0',
  `COLUMN_DEFAULT` longtext,
  `IS_NULLABLE` varchar(3) NOT NULL DEFAULT '',
  `DATA_TYPE` varchar(64) NOT NULL DEFAULT '',
  `CHARACTER_MAXIMUM_LENGTH` bigint(21) unsigned DEFAULT NULL,
  `CHARACTER_OCTET_LENGTH` bigint(21) unsigned DEFAULT NULL,
  `NUMERIC_PRECISION` bigint(21) unsigned DEFAULT NULL,
  `NUMERIC_SCALE` bigint(21) unsigned DEFAULT NULL,
  `DATETIME_PRECISION` bigint(21) unsigned DEFAULT NULL,
  `CHARACTER_SET_NAME` varchar(32) DEFAULT NULL,
  `COLLATION_NAME` varchar(32) DEFAULT NULL,
  `COLUMN_TYPE` longtext NOT NULL,
  `COLUMN_KEY` varchar(3) NOT NULL DEFAULT '',
  `EXTRA` varchar(30) NOT NULL DEFAULT '',
  `PRIVILEGES` varchar(80) NOT NULL DEFAULT '',
  `COLUMN_COMMENT` varchar(1024) NOT NULL DEFAULT ''
) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```

**8.0.22**
```sql
CREATE ALGORITHM=UNDEFINED DEFINER=`mysql.infoschema`@`localhost` SQL SECURITY DEFINER VIEW `COLUMNS` AS
 select (`cat`.`name` collate utf8_tolower_ci) AS
 `TABLE_CATALOG`,(`sch`.`name` collate utf8_tolower_ci) AS
 `TABLE_SCHEMA`,(`tbl`.`name` collate utf8_tolower_ci) AS
 `TABLE_NAME`,(`col`.`name` collate utf8_tolower_ci) AS
 `COLUMN_NAME`,`col`.`ordinal_position` AS
 `ORDINAL_POSITION`,`col`.`default_value_utf8` AS
 `COLUMN_DEFAULT`,if((`col`.`is_nullable` = 1),'YES','NO') AS
 `IS_NULLABLE`,substring_index(substring_index(`col`.`column_type_utf8`,'(',1),' ',1) AS
 `DATA_TYPE`,internal_dd_char_length(`col`.`type`,`col`.`char_length`,`coll`.`name`,0) AS
 `CHARACTER_MAXIMUM_LENGTH`,internal_dd_char_length(`col`.`type`,`col`.`char_length`,`coll`.`name`,1) AS
 `CHARACTER_OCTET_LENGTH`,if((`col`.`numeric_precision` = 0),NULL,`col`.`numeric_precision`) AS
 `NUMERIC_PRECISION`,if(((`col`.`numeric_scale` = 0) and (`col`.`numeric_precision` = 0)),NULL,`col`.`numeric_scale`) AS
 `NUMERIC_SCALE`,`col`.`datetime_precision` AS
 `DATETIME_PRECISION`,(case `col`.`type` when 'MYSQL_TYPE_STRING' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_VAR_STRING' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_VARCHAR' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_TINY_BLOB' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_MEDIUM_BLOB' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_BLOB' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_LONG_BLOB' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_ENUM' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) when 'MYSQL_TYPE_SET' 
 	then if((`cs`.`name` = 'binary'),NULL,`cs`.`name`) 
 	else NULL end) AS
 `CHARACTER_SET_NAME`,(case `col`.`type` when 'MYSQL_TYPE_STRING' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_VAR_STRING'
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_VARCHAR' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_TINY_BLOB' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_MEDIUM_BLOB' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_BLOB' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_LONG_BLOB' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_ENUM' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) when 'MYSQL_TYPE_SET' 
 	 then if((`cs`.`name` = 'binary'),NULL,`coll`.`name`) 
 	 else NULL end) AS
 `COLLATION_NAME`,`col`.`column_type_utf8` AS
 `COLUMN_TYPE`,`col`.`column_key` AS
 `COLUMN_KEY`,internal_get_dd_column_extra((`col`.`generation_expression_utf8` is null),`col`.`is_virtual`,`col`.`is_auto_increment`,`col`.`update_option`,if(length(`col`.`default_option`),true,false),`col`.`options`) AS
 `EXTRA`,get_dd_column_privileges(`sch`.`name`,`tbl`.`name`,`col`.`name`) AS
 `PRIVILEGES`,ifnull(`col`.`comment`,'') AS
 `COLUMN_COMMENT`,ifnull(`col`.`generation_expression_utf8`,'') AS
 `GENERATION_EXPRESSION`,`col`.`srs_id` AS
 `SRS_ID` from (((((`mysql`.`columns` `col` 
 	join `mysql`.`tables` `tbl` on((`col`.`table_id` = `tbl`.`id`))) 
 	join `mysql`.`schemata` `sch` on((`tbl`.`schema_id` = `sch`.`id`))) 
 	join `mysql`.`catalogs` `cat` on((`cat`.`id` = `sch`.`catalog_id`))) 
 	join `mysql`.`collations` `coll` on((`col`.`collation_id` = `coll`.`id`))) 
 	join `mysql`.`character_sets` `cs` on((`coll`.`character_set_id` = `cs`.`id`))) 
 where ((0 <> internal_get_view_warning_or_error(`sch`.`name`,`tbl`.`name`,`tbl`.`type`,`tbl`.`options`)) 
 	and (0 <> can_access_column(`sch`.`name`,`tbl`.`name`,`col`.`name`)) 
 	and (0 <> is_visible_dd_object(`tbl`.`hidden`,(`col`.`hidden` <> 'Visible'))));

```

And then this method cannot handle the input:

https://github.com/Sequel-Ace/Sequel-Ace/blob/8a2dbe4d5a9483398ef7c9338f370997c3dd8a1d/Source/Controllers/DataControllers/SPTableData.m#L812-L820

@Sequel-Ace/all Anyone else got time to look at this one?

**Comment by jamesstout** [comment]
I can make it not crash, but then the content of `COLUMNS` is gibberish.

**Comment by lphilps** [comment]
Thanks for looking into this James.  What a bizarre scenario, who would have expected a VIEW there?

Oddly enough, I originally went to information_schema.TABLES yesterday to get a list of all the views in a particular database. While Sequel Ace crashed in the Contents view, entering this in the query editor worked "just fine"

```
SELECT * FROM `information_schema`.`TABLES` WHERE `TABLE_SCHEMA` = 'dbName' AND `TABLE_TYPE` = 'VIEW';
```

Not sure if that observation is useful in any way, but ...

**Comment by jamesstout** [comment]
> Not sure if that observation is useful in any way, but ...

It's given me an idea for a fix ..  thanks.

**Comment by jamesstout** [comment]
@Sequel-Ace/all What do you think. It's a bit of a hack?

If `SHOW CREATE TABLE X` returns a string containing `CREATE ALGORITHM` and `DEFINER VIEW`, it means that create table is returning a view, which may or may not be parseable.

In that case we can create a temp table in the `mysql` db as, using COLUMNS as an example:

```sql
CREATE TEMPORARY TABLE `mysql`.`my_temp_table` SELECT * FROM `information_schema`.`COLUMNS`
```

Then we call 
```objc
[self informationForTable:@"my_temp_table" fromDatabase:@"mysql"]
```

and as it's now a table, it works fine.

Then at the end:
```sql
DROP TABLE `mysql`.`my_temp_table`
```

We can use this exclusively for Mysql 8+ and table = information_schema.

One issue is the Table Info screen doesn't have all the data to fill all the fields. But that's a small issue I think.

**Comment by Kaspik** [comment]
I feel like we need to treat MySQL 8 differently thus have different parsing method rather than workarounds - because it will stay as the new way.

So `informationForTable` should accept mysql version (enum?) and then decide based on that?

Here is a topic about it with link to a blog post - https://stackoverflow.com/a/54541147/4927614

**Comment by jamesstout** [comment]
Thanks. Will take a look. I've implemented the [temp table workaround anyway](https://github.com/Sequel-Ace/Sequel-Ace/tree/fix-information-schema-crash-883-temp-table) (I was doing it while waiting for opinions)

**Comment by jamesstout** [comment]
The problem is, table data is produced by this call:
```objc
NSDictionary *tableData = [self parseCreateStatement:tableCreateSyntax ofType:[resultFieldNames objectAtIndex:0]];
```
With `tableCreateSyntax` being the result from calling `SHOW CREATE TABLE X`.

But then check the comment for `parseCreateStatement`:

```objc
/**
* This function is rather long due to the painful parsing required, but is fast.
 *
 * *WARNING* This method is only designed to handle the output of a "SHOW CREATE ..." query.
 *  DO NOT try to use it with user-defined input. **The code does not handle the full possible syntax!**
 */
- (NSDictionary *)parseCreateStatement:(NSString *)tableDef ofType:(NSString *)tableType
```

We need to fix parseCreateStatement to handle this view syntax.

**Comment by jamesstout** [comment]
Reading all that led me to [this](https://dev.mysql.com/doc/refman/8.0/en/show-columns.html):

> `SHOW COLUMNS` displays information about the columns in a given table. **It also works for views**. 

The output is the same for < 8.0.0 and > 8.0.0 ...

**Comment by jamesstout** [comment]
Updated branch with a SHOW COLUMNS method rather than a temp table.

https://github.com/Sequel-Ace/Sequel-Ace/tree/fix-information-schema-crash-883-temp-table

Need to test that it works properly, then test for < 8.0.0. Then, maybe we can use one method for all.

**Comment by Jason-Morcos** [comment]
> Updated branch with a SHOW COLUMNS method rather than a temp table.
> 
> https://github.com/Sequel-Ace/Sequel-Ace/tree/fix-information-schema-crash-883-temp-table
> 
> Need to test that it works properly, then test for < 8.0.0. Then, maybe we can use one method for all.

Tested your fix and it definitely works with MySQL 5.7, @jamesstout! :D

**Comment by jamesstout** [comment]
I think I've managed to create a version that uses the same code for all server versions. Will PR for testing.
